### PR TITLE
Extract highly serviced train stations

### LINKED ISSUE #294: Parse and process train time table data
Built on discussion in #256 we can now access rail timetable data from [atoc](https://data.atoc.org/data-download) and we have a [list of stations](https://github.com/jwestw/UK-Train-Station-Locations/blob/master/uk-train-stations-dictonary.json), their CRS codes and their coordinates.

This ticket sets out the subtasks required to parse and clean the data, and calculate which stops are highly serviced.

- [x] #293
- [x] #299 
- [x] #295 
- [ ]  #328

### LINKED ISSUE #328: Filter train timetable data and create highly serviced stops
Reusing alot of code for bus timetable data

- Filter trains to only hours of interest
- Extract stops for a given day
- Create frequency dataframe for stops
- Export highly serviced train stops.

### PR DESCRIPTION
**Pull request by nkshaw23**
Code to extract highly serviced train stations. Follows same logic as bus timetable module. Main changes are:

1) Code structure changed to group loading tools and cleaning tools for each dataset (`mca` and `msn`). Improves readability and removes duplicated code.
2) Days of the week in schedule data extracted as `int`
3) `filter_by_day` function in data `transform` has been generalised as it is used across both bus and train scripts.
4) Easting and Northing added to `station_locations.csv`. New file uploaded to sync.
5) Output file names updated for bus timetable process to allow distinction between buses and trains. Files updated on sync.
6) Code added to filter timetable by a day of the week, create frequency dataframe and extract highly serviced stops.

Closes #294 
Closes #328

### CODE REVIEW

File: config.yaml [hunk 1]
```diff
@@ -38,12 +38,15 @@ ENG_bus_timetable_data: 'https://data.bus-data.dft.gov.uk/timetable/download/gtf
 auto_download_bus: true
 early_bus_hour: 06
 late_bus_hour: 20
+early_train_hour: 06
```
Comment (jwestw) [inline config.yaml:41]: I wonder if we need the train and bus times separate here. Wouldn't we always filter the same time for both?

We could have a setting like:
"late_timetable_hour" 

or 

"late_transport_hour"


What do you think?

File: config.yaml [hunk 2]
```diff
@@ -38,12 +38,15 @@ ENG_bus_timetable_data: 'https://data.bus-data.dft.gov.uk/timetable/download/gtf
...
+late_train_hour: 20
 high_cap_buffer: 1000
 low_cap_buffer: 500
 timetable_day: 'wednesday'
 day_filter: 'general' #exact
 train_msn_filename: 'ttisf467.msn'
 train_mca_filename: 'ttisf467.mca'
+station_locations: 'station_locations.csv'
```
Comment (jwestw) [inline config.yaml:49]: We need to add something in the docs about where to get this from (that repo I set up). 

https://github.com/jwestw/UK-Train-Station-Locations

**Review by jwestw — COMMENTED** [review]
I can't pull and run the code at the moment because I am on an ONS computer, but there's a couple of comments for you to consider before I do. :)

File: config.yaml (hunk 1 above)
Comment (nkshaw23) [reply config.yaml:41]: Agree - wasnt sure so left seperate but might as well combine.

File: config.yaml (hunk 2 above)
Comment (nkshaw23) [reply config.yaml:49]: Already added

**Comment by james-westwood** [comment]
I have included .gitkeep files so the data folder structure is preserved. 

[gitkeeps to preserve folder structure](https://github.com/ONSdigital/SDG_11.2.1/pull/329/commits/7b82aada1d308001348a6ebde51f46f86f4fe096)

**Comment by james-westwood** [comment]
I am getting a warning

```
/home/james/programming/SDG_11.2.1/SDG_train_timetable.py:272: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  serviced_train_stops_df['departure_time'] = serviced_train_stops_df['departure_time'].str.slice(0,2)
```

@nkshaw23 Should we look into this?

**Comment by james-westwood** [comment]
The script runs and the df looks good. 

image

**Review by james-westwood — APPROVED** [review]
This all looks good! Happy to approve

**Pull request merged**